File: Interview/DS/Sorting/Bubble_sort/bubble_sort.py
'''
NOTES:
Time Complexity: O(N^2)
Space Complexity: O(1) [temp for swapping]
'''

def bubble_sort(arr):
    n = len(arr)
    
    # Outer loop runs n times
    for i in range(n):
        
        # Inner loop runs n-i-1 times
        for j in range(0, n-i-1):
            
            # If the current element is greater than the next element, swap them
            if arr[j] > arr[j+1]:
                arr[j], arr[j+1] = arr[j+1], arr[j]
                
    # Return the sorted list
    return arr

# Example usage
arr = [64, 34, 25, 12, 22, 11, 90]
print("Before sorting: ", arr)
arr = bubble_sort(arr)
print("After sorting: ", arr)

File: Interview/DS/Sorting/Bubble_sort/bubble_sort.java
import java.util.Arrays;

public class BubbleSort {
    public static void main(String[] args) {
        int[] arr = {5, 1, 4, 2, 8};
        bubbleSort(arr);
        System.out.println("Sorted array: " + Arrays.toString(arr));
    }

    public static void bubbleSort(int[] arr) {
        int n = arr.length;
        
        // outer loop: iterates n-1 times
        for (int i = 0; i < n-1; i++) {
            
            // inner loop: iterates n-i-1 times
            for (int j = 0; j < n-i-1; j++) {
                
                // compare adjacent elements and swap if in wrong order
                if (arr[j] > arr[j+1]) {
                    int temp = arr[j];
                    arr[j] = arr[j+1];
                    arr[j+1] = temp;
                }
            }
        }
    }
}
